java/src/main/java/com/ibm/cloud/sdk/authentication/Authenticator.java

```diff
@@ -0,0 +1,276 @@
+/* -----------------------------------------------------------------------------
+ * Copyright IBM Corp. 2018
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ----------------------------------------------------------------------------- */
+
+package com.ibm.cloud.sdk.authentication;
+
+import java.io.StringReader;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+
+import javax.json.Json;
+import javax.json.JsonObject;
+import javax.ws.rs.client.Client;
+import javax.ws.rs.client.ClientBuilder;
+import javax.ws.rs.client.Entity;
+import javax.ws.rs.client.Invocation;
+import javax.ws.rs.core.Form;
+import javax.ws.rs.core.Response;
+
+import com.ibm.cloud.sdk.exceptions.AuthenticationError;
+import com.ibm.cloud.sdk.impl.AuthenticationResult;
+import com.ibm.cloud.sdk.util.Communication;
+
+public class Authenticator {
+
+    public enum GrantType {
+        apiKey
+    };
+
+    final private String clientId;
+    final private String clientSecret;
+    final private GrantType grantType;
+    final private String apiKey;
+    final private boolean requestUAATokens;
+    final private String uaaClientId;
+    final private String uaaClientSecret;
+    final private String uaaRedirectUri;
+    final private boolean requestIMSPortalToken;
+    final private String accountId;
+
+    private Authenticator(String clientId, String clientSecret, GrantType grantType, String apiKey, boolean requestUAATokens, String uaaClientId, String uaaClientSecret,
+            String uaaRedirectUri, boolean requestIMSPortalToken, String accountId) {
+        this.clientId = clientId;
+        this.clientSecret = clientSecret;
+        this.grantType = grantType;
+        this.apiKey = apiKey;
+        this.requestUAATokens = requestUAATokens;
+        this.uaaClientId = uaaClientId;
+        this.uaaClientSecret = uaaClientSecret;
+        this.uaaRedirectUri = uaaRedirectUri;
+        this.requestIMSPortalToken = requestIMSPortalToken;
+        this.accountId = accountId;
+    }
+
```

Comment:
Wouldn't it make sense to offer some convenience creators for the typical Authenticator usages on top of the builder pattern currently exposed? I.e. like `public static Authenticator newApiKeyAuthenticator(String apiKey) {
	            return newBuilder().useApiKey(apiKey).build();
	        }`

Reply:
If we provide something more simple, I would see this as a method on IBMCloudClient, so that the user does not need to handle with the `Authenticator` at all, e.g. `authenticateUsingApiKey(String apiKey)`.

I think the Authenticator should only be used with the Builder model if this level of granularity is reallly necessary...

Reply:
Yes, that sounds good too. Let's do that.

Reply:
next round... added the changes